Add new `duplicated_attributes` lint

GuillaumeGomez opened the pull request:
It's a lint idea that @llogiq gave me while reviewing another PR.

There are some limitations, in particular for the "output". Initially I wanted to make it possible for directly lint against the whole attribute if its parts were all duplicated, but then I realized that the output would be chaotic if the duplicates were coming from different attributes, so I preferred to go to the simplest way and simply emit a warning for each entry. Not the best, but makes the implementation much easier.

Another limitation is that `cfg_attr` would be a bit more tricky to implement because we need to check if two `cfg` sets are exactly the same. I added a FIXME and will likely come back to it later.

And finally, I updated the `cargo dev update_lints` command because the generated `tests/ui/rename.rs` file was emitting the `duplicated_attributes` lint, so I allowed this lint inside it to prevent it from working.

changelog: Add new `duplicated_attributes` lint

GuillaumeGomez commented:
Manish said they're busy so let's pick someone else.

r? @blyxyas

bors commented:
:umbrella: The latest upstream changes (presumably #12354) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

blyxyas reviewed (changes requested):
I've done some tinkering with that `FIXME` and `cfg_attr`, I think it wouldn't be too hard to support it.

Really good first iteration! :3

GuillaumeGomez commented on clippy_dev/src/update_lints.rs:
```diff
@@ -689,6 +689,8 @@ fn gen_deprecated_lints_test(lints: &[DeprecatedLint]) -> String {
     let mut seen_lints = HashSet::new();
     let mut res: String = GENERATED_FILE_COMMENT.into();
+
+    res.push_str("#![allow(clippy::duplicated_attr)]\n");
```
The `uibless` run kept failing because this new lint was emitted, making it fail.

GuillaumeGomez commented on clippy_lints/src/attrs.rs:
```diff
@@ -512,6 +515,31 @@ declare_clippy_lint! {
+    /// Checks if an item has duplicated attributes.
+    ///
+    /// ### Why is this bad?
+    /// More code for no reason.
```
Much better indeed. ^^'

GuillaumeGomez commented on clippy_lints/src/attrs.rs:
```diff
@@ -512,6 +515,31 @@ declare_clippy_lint! {
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks if an item has duplicated attributes.
```
:+1:

GuillaumeGomez commented:
Applied suggestions.

bors commented:
:umbrella: The latest upstream changes (presumably #12394) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
Renamed the lint as suggested, fixed merge conflict and double checked: still doesn't work without the `rename.rs` generation update.

GuillaumeGomez commented:
And even fixed CI. :D

GuillaumeGomez commented:
Just checked, still not working. Giving the steps I'm doing to reproduce:

```terminal
$ git rebase -i HEAD~4 # I remove the commit making the change
Successfully rebased and updated detached HEAD.
$ cargo dev update_lints
   Compiling clippy_dev v0.0.1 (/home/imperio/rust/clippy/clippy_dev)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 2.68s
     Running `target/debug/clippy_dev update_lints`
$ TESTNAME=rename.rs cargo uitest
   Compiling clippy_config v0.1.78 (/home/imperio/rust/clippy/clippy_config)
   Compiling clippy_utils v0.1.78 (/home/imperio/rust/clippy/clippy_utils)
   Compiling clippy_lints v0.1.78 (/home/imperio/rust/clippy/clippy_lints)
   Compiling clippy v0.1.78 (/home/imperio/rust/clippy)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 35.57s
     Running tests/compile-test.rs (target/debug/deps/compile_test-b0bc7a5dfdc2e878)
Building dependencies ... ok
tests/ui/rename.rs ... FAILED                                                                           

FAILED TEST: tests/ui/rename.rs
command: CLIPPY_CONF_DIR="tests" "/home/imperio/rust/clippy/target/debug/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/home/imperio/rust/clippy/target/debug/deps" "--extern=clippy_config=/home/imperio/rust/clippy/target/debug/deps/libclippy_config-5d14e48a50025dbd.rlib" "--extern=clippy_lints=/home/imperio/rust/clippy/target/debug/deps/libclippy_lints-992d72a6c56735a3.rlib" "--extern=clippy_utils=/home/imperio/rust/clippy/target/debug/deps/libclippy_utils-da6e3f676bd0803e.rlib" "--extern=futures=/home/imperio/rust/clippy/target/debug/deps/libfutures-b2e68350a4947d59.rlib" "--extern=if_chain=/home/imperio/rust/clippy/target/debug/deps/libif_chain-0d767436d7e18eee.rlib" "--extern=itertools=/home/imperio/rust/clippy/target/debug/deps/libitertools-dcd2b845627dc871.rlib" "--extern=parking_lot=/home/imperio/rust/clippy/target/debug/deps/libparking_lot-7b8f849c7fda2d42.rlib" "--extern=quote=/home/imperio/rust/clippy/target/debug/deps/libquote-28a9e8d5e247e24e.rlib" "--extern=regex=/home/imperio/rust/clippy/target/debug/deps/libregex-66a489862c42a796.rlib" "--extern=serde=/home/imperio/rust/clippy/target/debug/deps/libserde-30375b3bb2f254b9.rlib" "--extern=serde_derive=/home/imperio/rust/clippy/target/debug/deps/libserde_derive-68eaa659e149f4ae.so" "--extern=syn=/home/imperio/rust/clippy/target/debug/deps/libsyn-251cd79fb57cdb56.rlib" "--extern=tokio=/home/imperio/rust/clippy/target/debug/deps/libtokio-78db72a98dbeacfc.rlib" "--out-dir" "target/ui_test/tests/ui" "tests/ui/rename.fixed" "--edition" "2021" "--crate-name" "rename"

error: rustfix failed with exit status: 1

full stderr:
error: duplicated attribute
  --> tests/ui/rename.fixed:96:9
   |
LL | #![warn(for_loops_over_fallibles)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first defined here
  --> tests/ui/rename.fixed:95:9
   |
LL | #![warn(for_loops_over_fallibles)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
help: remove this attribute
  --> tests/ui/rename.fixed:96:9
   |
LL | #![warn(for_loops_over_fallibles)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
   = note: `-D clippy::duplicated-attributes` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::duplicated_attributes)]`

error: duplicated attribute
  --> tests/ui/rename.fixed:97:9
   |
LL | #![warn(for_loops_over_fallibles)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first defined here
  --> tests/ui/rename.fixed:95:9
   |
LL | #![warn(for_loops_over_fallibles)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
help: remove this attribute
  --> tests/ui/rename.fixed:97:9
   |
LL | #![warn(for_loops_over_fallibles)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors


full stdout:


FAILURES:
    tests/ui/rename.rs

test result: FAIL. 1 failed; 966 filtered out;

thread 'main' panicked at tests/compile-test.rs:176:6:
called `Result::unwrap()` on an `Err` value: tests failed

Location:
    /home/imperio/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ui_test-0.22.2/src/lib.rs:395:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: test failed, to rerun pass `--test compile-test`
```

As you can see, it fails. So no, I confirm the change is needed.

GuillaumeGomez commented:
> and then got lost in playing a little bit of Helldivers 2, I liked it.

Having same issue with baldur's gate 3. :D

> Maybe an artifact of previous versions? If there's black magic and on some platforms that forced attribute is necessary, we'll just keep it there, no big deal :)

I rebased on the latest clippy version (locally) and ran the commands from my previous message, still not working as you can see.

blyxyas commented:
Meow meow meow
@bors r+

bors commented:
:pushpin: Commit ae52a9d428209894a67488d67458ce7e0031b2df has been approved by `blyxyas`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=blyxyas ae52a9d428209894a67488d67458ce7e0031b2df -->
<!-- homu: {"type":"Approved","sha":"ae52a9d428209894a67488d67458ce7e0031b2df","approver":"blyxyas","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit ae52a9d428209894a67488d67458ce7e0031b2df with merge f685a4b3f1533091fd44580ad3d4aadba033852e...
<!-- homu: {"type":"BuildStarted","head_sha":"ae52a9d428209894a67488d67458ce7e0031b2df","merge_sha":"f685a4b3f1533091fd44580ad3d4aadba033852e"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/8238052520/job/22528128599), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/8238052523/job/22528135910), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/8238052517/job/22528328285)
Approved by: blyxyas
Pushing f685a4b3f1533091fd44580ad3d4aadba033852e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"blyxyas","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/8238052517/job/22528328285","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/8238052520/job/22528128599","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/8238052523/job/22528135910"},"merge_sha":"f685a4b3f1533091fd44580ad3d4aadba033852e"} -->
